sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -209,6 +209,19 @@ case class AdaptiveSparkPlanExec(
 
   override def output: Seq[Attribute] = inputPlan.output
 
+  // Try our best to give a stable output partitioning and ordering.
```

Comment:
I'm trying to understand this "best effort". AFAIK, table cache is lazy. For a query that accesses a cached query the first time, the cached query is not executed yet so we don't know the output partitioning/ordering and can't optimize out shuffles. But when the cached query is accessed the next time, it's already executed and we know the output partitioning/ordering.

Reply:
yes, in general the first action for a cached plan is `count`, e.g. `CacheTableAsSelectExec`, so I think it is a not big  issue that we can not optimize the shuffle/sort for the first action.

The usage of the cache is: user wants to reference it multi-times, then this optimization will help a lot.

Reply:
This would be super limited use... and cause inconsistency.

I'd only return output partitioning if there is a user repartition op in the end. In other words, only if AQE plan is required to preserve user specified partitioning.

Reply:
Unfortunately we hint this.. per my experience, user always caches an arbitrary df and use the cached df to build an another arbitrary df. So why can't we preserve the partitioning/ordering of the cached plan ? If you really feel inconsistency in AdaptiveSparkPlanExec, we can probably move to InMemoryRelationExec.

My original idea is to do the both but feel a little overkill (requiredOrdering should be inferred separately like https://github.com/apache/spark/pull/35924)
```scala
requiredDistribution.map(_.createPartitioning(conf.shufflePartitions)).getOrElse {
  if (isFinalPlan) {
    executedPlan.outputPartitioning
  } else {
    super.outputPartitioning
  }
}
```
A useful distribution before caching is few in production since `repartition(col)` will introduce skew